luatest/replica_set.lua

```diff
@@ -0,0 +1,231 @@
+            :format(alias))
+    end
+end
+
+--- Start all servers in the replica set.
+-- Optionally waits until all servers are ready.
+--
+-- @tab[opt] opts
+-- @bool[opt] opts.wait_until_ready Wait until all servers are ready.
+--   Defaults to `true`.
+function ReplicaSet:start(opts)
+    checks('table', {wait_until_ready = '?boolean'})
+
+    for _, server in ipairs(self.servers) do
+        if not server.process then
+            server:start({wait_until_ready = false})
+        end
+    end
+
+    local wait_until_ready = true -- the option is `true` by default
+    if opts ~= nil and opts.wait_until_ready ~= nil then
+        wait_until_ready = opts.wait_until_ready
+    end
+    if wait_until_ready then
+        for _, server in ipairs(self.servers) do
+            server:wait_until_ready()
+        end
+    end
+end
+
+--- Stop all servers in the replica set.
+function ReplicaSet:stop()
+    for _, server in ipairs(self.servers) do
+        server:stop()
+    end
+end
+
+--- Clean working directories of all servers in the replica set.
+-- Should be invoked only for a stopped replica set.
+function ReplicaSet:clean()
+    for _, server in ipairs(self.servers) do
+        server:clean()
+    end
+end
+
+--- Stop all servers in the replica set and clean their working directories.
+function ReplicaSet:drop()
+    for _, server in ipairs(self.servers) do
+        server:stop()
+        server:clean()
+    end
+end
+
+--- Get a server which is a writable node in the replica set.
+function ReplicaSet:get_leader()
+    for _, server in ipairs(self.servers) do
+        if server:exec(function() return box.info.ro end) == false then
+            return server
+        end
+    end
+end
+
+--- Wait until every node is connected to every other node in the replica set.
+--
+-- @tab[opt] opts
+-- @number[opt] opts.timeout Timeout in seconds to wait for full mesh.
+--   Defaults to 60.
+-- @number[opt] opts.delay Delay in seconds between attempts to check full mesh.
+--   Defaults to 0.1.
+function ReplicaSet:wait_for_fullmesh(opts)
+    checks('table', {timeout = '?number', delay = '?number'})
+    if not opts then opts = {} end
+    local config = {timeout = opts.timeout or 60, delay = opts.delay or 0.1}
+    helpers.retrying(config, function(replica_set)
+        for _, server1 in ipairs(replica_set.servers) do
+            for _, server2 in ipairs(replica_set.servers) do
+                if server1 ~= server2 then
+                    local server1_id = server1:get_instance_id()
+                    local server2_id = server2:get_instance_id()
+                    if server1_id ~= server2_id then
+                        server1:assert_follows_upstream(server2_id)
+                    else
+                        -- If IDs are equal, nodes are anonymous replicas and
+                        -- not registered yet. Raise an error to retry checking
+                        -- full mesh again.
```

Comment:
So if we found anonymous replicas, we start the search again from the start? Could we wait and retry on this particular pair and then move on? Or is it ineffective?

Reply:
Probably we could. But I wouldn't like to introduce the new logic for now.